TKG(1.5)-airgap on aws Reference Design and Deployment Guide (#180)
vmware-tanzu-labs/tanzu-validated-solutions: src/deployment-guides/tko-aws-airgap.md

# Deploy Tanzu for Kubernetes Operations on AWS in an Airgapped Environment

This document outlines the steps for deploying VMware Tanzu for Kubernetes Operations on AWS in an airgapped (internet-restricted) environment. The deployment is based on the reference design provided in [VMware Tanzu Kubernetes Grid on AWS Airgap Reference Design](../reference-designs/tko-on-aws-airgap.md).

## Prerequisites
Before deploying VMware Tanzu for Kubernetes Operations in an AWS airgapped environment, ensure that the following are set up.

* **AWS Account**: An IAM user account with **administrative privileges**.

> * **AWS Resource Quotas**: Sufficient quotas to support both the management cluster and the workload clusters in your deployment. Otherwise, the cluster deployments will fail. Depending on the number of workload clusters you plan to deploy, you may need to increase the AWS services quotas from their default values. You will need to increase the quota in every region in which you deploy Tanzu Kubernetes Grid.

Reviewer: ```suggestion
* **AWS Resource Quotas:** Sufficient quotas to support both the management cluster and the workload clusters in your deployment. Otherwise, the cluster deployments will fail. Depending on the number of workload clusters you plan to deploy, you may need to increase the AWS services quotas from their default values. You will need to increase the quota in every region in which you deploy Tanzu Kubernetes Grid.
```

	* [AWS service quotas](https://docs.aws.amazon.com/general/latest/gr/aws_service_limits.html) in the AWS.

* **An Internet-connected Linux bootstrap machine**
The bootstrap machine can be a local device such as a laptop or a virtual machine running in, for example, VMware Workstation or Fusion. You will use the bootstrap machine to create the AWS VPC and jumpbox. The bootstrap machine:
	* Is not inside the internet-restricted environment or is able to access the domains listed in Proxy Server Allowlist.
	* Has the Docker client app installed.
	* Has [imgpkg](https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.5/vmware-tanzu-kubernetes-grid-15/GUID-install-cli.html) installed.
	* Has the latest version of yq installed.
	* Has the latest version of jq installed.
	* Has AWS CLI installed.
	* Has the [Carvel Tools](https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.5/vmware-tanzu-kubernetes-grid-15/GUID-install-cli.html) installed, if you intend to install one or more of the optional packages provided by Tanzu Kubernetes Grid, such as Harbor.

> * **VMware Cloud**: Access to [VMware Cloud]( https://customerconnect.vmware.com/login ) to download Tanzu CLI.

Reviewer: ```suggestion
* **VMware Cloud:** Access to [VMware Cloud]( https://customerconnect.vmware.com/login ) to download Tanzu CLI.
```

## Overview of the Deployment Steps

The main steps to deploy Tanzu for Kubernetes Operations on AWS EC2 are as follows. Each step links to more detailed instructions.

1. [Set up AWS Infrastructure](#aws-infra).
2. [Creating offline JumpBox](#offline-jumpbox)
3. [Create and Set Up a Private container registry](#private-repo).
4. [Copy the container images required to deploy Tanzu Kubernetes Grid](#copy-tkg-img).  
5. [Prepare an Internet-Restricted Environment](#prepEnv).
6. [Install Tanzu Kubernetes Grid Management Cluster](#install-tkg).
7. [Examine the Management Cluster Deployment](#examine-cluster).
8. [Deploy Workload Clusters](#deploy-workload-cluster).
9. [Install and Configure Packages into Workload Clusters](#install-packages).
10. [Logs and Troubleshooting](#logs).
11. [Delete Clusters](#cluster-mgmt).
12. [Airgapped STIG/FIPS deployment on AWS](#stig-fips). 	
13. [Tanzu Kubernetes Grid Upgrade](#upgrade-tkg).

> ## <a id=aws-infra> </a> Set up AWS Infrastructure

Reviewer: ```suggestion
## <a id=aws-infra> </a> Set Up AWS Infrastructure
```

The following describes the steps to create your AWS environment and configure your network. Follow the steps in the order provided.

  ### Create an internet-restricted Virtual Private Cloud (VPC)
   Follow these steps to create an internet-restricted (offline) AWS VPC or refer to the [AWS VPC documentation](https://docs.aws.amazon.com/vpc/latest/userguide/working-with-vpcs.html). This offline VPC should not have a NAT or internet gateway. Create your offline VPC with private subnets only.
Steps to follow:

>  **Step 01. Create a VPC**

Reviewer: ```suggestion
 **Step 1. Create a VPC.**
```

Reviewer: Removed unnecessary leading zeros from these steps.

> * Log in to your AWS Console and select VPC service.

Reviewer: ```suggestion
* Log in to your AWS Console and select **VPC service**.
```

* Create your VPC with Valid CIDR and name.

> **Step 02.  Create 3 Private Subnets**

Reviewer: ```suggestion
**Step 2.  Create 3 private subnets.**
```

* Click **Subnet** and create your subnet with:
* Private Subnet 1, Private Subnet 2 and Private Subnet 3 valid Name & VPC.

> * Valid Subnet range which is valid IPv4 CIDR Block.

Reviewer: ```suggestion
* Valid Subnet range, which is a valid IPv4 CIDR block.
```

**Step 03. Create Private Route Table**

> * Create a Route table in the same VPC.

Reviewer: ```suggestion
* Create a route table in the same VPC.
```

> * Make sure you selected the right VPC and gave a proper tag.

Reviewer: ```suggestion
* Make sure you selected the right VPC and gave it a proper tag.
```

> **Step 04. Add Private Subnet in Private Route Table**

Reviewer: ```suggestion
**Step 4. Add a private subnet in the private route table.**
```

> * Edit the Subnet Association.

Reviewer: ```suggestion
* Edit the subnet association.
```

> * Select the PrivateSubnet checkbox.

Reviewer: ```suggestion
* Select the **PrivateSubnet** checkbox.
```

* Click **Save**.

> **Step 05. Edit DNS Resolution and Hostname**

Reviewer: ```suggestion
**Step 5. Edit DNS Resolution and Hostname.**
```

* Click **Action** and Edit DNS hostname

> * Select DNS Hostname and click **Save**.

Reviewer: ```suggestion
* Select **DNS Hostname** and click **Save**.
```

> **Note:** If you are creating multiple offline VPCs, follow these instructions to create an [aws transit gateway](https://docs.aws.amazon.com/vpc/latest/tgw/tgw-getting-started.html).

Reviewer: ```suggestion
**Note:** If you are creating multiple offline VPCs, follow these instructions to create an [AWS transit gateway](https://docs.aws.amazon.com/vpc/latest/tgw/tgw-getting-started.html).
```

### Add VPC endpoints into offline VPC

> After you create the offline VPC, you must add the following endpoints to it. (VPC endpoints enable private connections between your VPC and supported AWS services.)

Reviewer: ```suggestion
After you create the offline VPC, you must add the following endpoints to it. VPC endpoints enable private connections between your VPC and supported AWS services.
```

**Service endpoints:**
* sts
* ssm
* ec2
* ec2messages
* elasticloadbalancing
* secretsmanager
* ssmmessages
* s3 (gateway type)

> See [aws create endpoints service documentation](https://docs.aws.amazon.com/vpc/latest/privatelink/create-endpoint-service.html) for further details.

Reviewer: ```suggestion
See the [AWS documentation for creating an endpoint service](https://docs.aws.amazon.com/vpc/latest/privatelink/create-endpoint-service.html) for further details.
```

> ## <a id=offline-jumpbox> </a> Create an offline JumpBox

Reviewer: ```suggestion
## <a id=offline-jumpbox> </a> Create an Offline Jumpbox
```

## <a id=prepEnv> </a> Prepare an Internet-Restricted Environment

Before you can deploy management clusters and Tanzu Kubernetes clusters in an internet-restricted environment, you need to prepare the environment. Follow the instructions in [Prepare an Internet-Restricted Environment](https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.5/vmware-tanzu-kubernetes-grid-15/GUID-mgmt-clusters-airgapped-environments.html).

## <a id=install-tkg></a> Deploy a Tanzu Kubernetes Grid Management Cluster

Create and edit a YAML configuration file. Then use the configuration file with CLI commands to deploy a management cluster.

This section describes how to deploy a Tanzu Kubernetes Grid management cluster from a configuration file using the Tanzu CLI.

Before creating a management cluster using the Tanzu CLI, define the base configuration for the cluster in a YAML file. Specify this file by using the `tanzu management-cluster create` command with the the `--file` option.

**Note** In the configuration file for the management cluster, enable the AWS internal load balancer as follows:

  `AWS_LOAD_BALANCER_SCHEME_INTERNAL: "true"`

> <_I added quotes around *true* - correct?_>

Reviewer: ```suggestion

```

Reviewer: Are the quote marks about **true** correct?

Using an internal load balancer scheme prevents the Kubernetes API server for the cluster from being accessed and routed over the internet. <_I reworded this note because the original seemed redundant. Is the meaning still correct?_>

To create a new Tanzu Kubernetes Grid management cluster, run the following command:

```bash
tanzu management-cluster create --file path/to/cluster-config-file.yaml
```

For more information about deploying a management cluster from a configuration file, see [Deploy Management Clusters from a Configuration File](https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.5/vmware-tanzu-kubernetes-grid-15/GUID-mgmt-clusters-deploy-cli.html).  

## <a id=examine-cluster></a> Examine the Management Cluster Deployment

> When the management cluster is deployed, either from the installer interface or from a configuration file using Tanzu CLI, Tanzu Kubernetes Grid uses a Kubernetes in Docker kind cluster on the jumpbox to create a temporary management cluster. kind is a tool for running Kubernetes clusters locally using Docker containers as Kubernetes nodes. <_ Correct? _>

Reviewer: ```suggestion
When the management cluster is deployed, either from the installer interface or from a configuration file using Tanzu CLI, Tanzu Kubernetes Grid uses a Kubernetes in Docker `kind` cluster on the jumpbox to create a temporary management cluster. `kind` is a tool for running Kubernetes clusters locally using Docker containers as Kubernetes nodes. 
```

After doing the network configuration, complete the steps described in this section to set up your jumpbox. You will download the Tanzu CLI to the jumpbox, which you will use to deploy the management cluster and workload clusters from the jumpbox. You also keep the Tanzu and Kubernetes configuration files for your deployments on your jumpbox.

1. Create a jumpbox.

	<!-- /* cSpell:disable */ -->

	```bash
   export vpcId=<offline vpc id>
   export WORKING_DIR=<local working dir>

	aws ec2 create-security-group --group-name "jumpbox-ssh" --description "To Jumpbox" --vpc-id "$vpcId" --output json > $WORKING_DIR/sg_jumpbox_ssh
	aws ec2 create-tags --resources $(jq -r .GroupId $WORKING_DIR/sg_jumpbox_ssh) --tags Key=Name,Value="jumpbox-ssh"
	# Allow ssh to jumpbox
	aws ec2 authorize-security-group-ingress --group-id  $(jq -r .GroupId $WORKING_DIR/sg_jumpbox_ssh) --protocol tcp --port 22 --cidr "0.0.0.0/0"

	# Save this file or use some team keypair already created
	aws ec2 create-key-pair --key-name tkg-kp --query 'KeyMaterial' --output text > tkgkp.pem
	chmod 400 tkgkp.pem

	# Find an AMI for your region https://cloud-images.ubuntu.com/locator/ec2/ (20.04)

> 	aws ec2 run-instances --image-id ami-036d46416a34a611c --count 1 --instance-type t2.xlarge --key-name tkg-kp --security-group-ids  $(jq -r .GroupId $WORKING_DIR/sg_jumpbox_ssh)   --subnet-id $(jq -r .Subnet.SubnetId $WORKING_DIR/subnet-pub-a)  --tag-specifications 'ResourceType=instance,Tags=[{Key=Name,Value=tkg-jumpbox}]' --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=64}' > $WORKING_DIR/instance_jb_starting

Reviewer 2: @sendjainabhi  I think we should refer to one of the private subnets which have been created above manually instead of reading via a file. 
$WORKING_DIR/subnet-pub-a
. What say?
cc @sunshetty

cc

Reviewer 3: I think the way we specify it is less error prone then telling them to read it manually

Author: @z4ce - there is no public subnet in airgap ra , and we need to create jumpbox within private subnet. i will have to change this in instructions.

Reviewer 2: i can see its env variable now, I just thought, as the user is creating the subnets manually. So he might not have subnet out file in working dir. But env variable works, I guess

	```
	<!-- /* cSpell:enable */ -->

2. Wait a few minutes for the instance to start. After it restarts, SSH to the jumpbox.

	<!-- /* cSpell:disable */ -->
	```bash
	aws ec2 describe-instances --instance-id $(jq -r '.Instances[0].InstanceId' $WORKING_DIR/instance_jb_starting) > $WORKING_DIR/instance_jb_started

> 	echo j IP: $(jq -r '.Reservations[0].Instances[0].PublicIpAddress' $WORKING_DIR/instance_jb_started)

Reviewer 2: @sendjainabhi As we are targetting offline VPC? shouldn't we use Jumbox's private address?

cc @sunshetty

Author: > @sendjainabhi As we are targetting offline VPC? shouldn't we use Jumbox's private address?
> 
> cc @sunshetty

fixed now.

Reviewer 2: (Y) thanks
